### ISSUE TITLE
flowtype "declare var module" causes transform-flow-strip-types to transpile module to _module

### ISSUE THREAD

**Issue by philiiiiiipp** [babel/babel#5718]
Choose one: is this a bug report or feature request?
**bug report**

When using flowtype's 
```js
declare var module
```
in a file which uses module.exports, transform-flow-strip-types produces
```js
_module.exports
```
instead of 
```js
module.exports
```

Here is the sample code to reproduce 
[https://github.com/philiiiiiipp/modules_test](https://github.com/philiiiiiipp/modules_test)

### Babel Configuration (.babelrc, package.json, cli command)
```js
"babel": {
    "presets": [
      "es2015",
      "stage-2"
    ],
    "plugins": [
      "transform-flow-strip-types"
    ]
  }
```

| software         | version(s)
| ---------------- | -------
| Babel            |  6.24.1 
| node             | v6.10.3
| npm              | 4.2.0
| Operating System | OSX 10.11.6

**Comment by joshwnj** [comment]
Hi @philiiiiiipp can you please elaborate on the purpose of `declare var module` here?  If I understand correctly this tells flow that `module` is a global: https://flow.org/en/docs/libdefs/creation/#toc-declaring-a-global-variable

Btw, from what I can see this is the place where `module` is being renamed as `_module`: https://github.com/babel/babel/blob/master/packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js#L158

**Comment by philiiiiiipp** [comment]
Yes exactly. I use this if I import a frontend library e.g. google maps which creates a global variable `google`.

In this case I used it to hide some flow errors while developing where module.hot.accept was not known by flow.

```
if (module.hot) {
  // Enable Webpack hot module replacement for reducers
  module.hot.accept(ROOT_REDUCER_PATH, () => {
    const nextRootReducer = require(ROOT_REDUCER_PATH);
    store.replaceReducer(nextRootReducer);
  });
}
```

**Comment by joshwnj** [comment]
@philiiiiiipp ah I see what you mean.

Could you try removing the `declare var module` from `main.js`, and instead add `flow-typed/module.js` like this?

```js
declare var module: {
  hot: {
    accept: (path: ?string, callback: ?() => void) => void
  }
}
```

That seems to satisfy a flow check, without causing the `module -> _module` rename.

**Comment by xtuc** [comment]
@joshwnj Thanks for your answer!

@philiiiiiipp If it's ok for you please close this issue.

**Comment by philiiiiiipp** [comment]
Sure I can do that, but isn't the _module behavior a semi critical bug? I mean this can silently break code.

**Comment by xtuc** [comment]
Yes, that's a good point.
We could maybe check if a built-in was re-declared and warn the user about it.

**Comment by philiiiiiipp** [comment]
Yes that would be one way, but I suppose it should actually be possible to do that without breaking things.

I am not quite familiar about babel works here, do the plugins `flow-strip-types` run before or after   the presets `es2015, stage-2, react` ? I guess at least `es2015` is running before `flow-strip-types`, then renames `module` and flow-strip-types does not pick up on it anymore.

**Comment by xtuc** [comment]
Yes, the presets run first. We currently have issues with plugin ordering https://github.com/babel/babel/pull/5735.

**Comment by philiiiiiipp** [comment]
Ah ok, in that case it might be a good idea to add a warning, (maybe even to crash ?) for the time being since as far as I see it there is no desirable outcome from the transpiled code other than potential bugs.

**Comment by LeonxLJX** [comment]
I'd like to take this one.

I've read through the report. Before changing anything I'd like to reproduce it against current `main` first — my reasoning is that a fix anchored to the actual observed behavior is much easier to review than one inferred from the report alone, especially if the surrounding code has moved since this was filed.

Plan: reproduce and confirm the root cause, then make the smallest change that addresses it (plus a regression test if the repo has coverage for this area). If the root cause turns out to live somewhere other than where the report points, I'll report back here before forcing a change.

Anything already known about this that I should read first?

(claiming via @LeonxLJX)
